src/backend/updates.rs

```diff
@@ -71,19 +78,93 @@ fn post_upgrade() {
 }
 
 async fn post_upgrade_fixtures() {
+    mutate(|state| state.read_only = true);
+    let total_posts = read(|state| state.posts.len());
+
+    // Init all stable structs becasue we do not need the old contents of the stable memory anymore
+    init_stable_structs();
+
+    loop {
+        let num_posts = read(|state| state.posts.len());
+        let _: Result<(), ()> = canisters::call_canister(id(), "migrate_posts", ((),))
+            .await
+            .map_err(|(code, err)| {
+                mutate(|state| {
+                    state.logger.error(format!(
+                        "couldn't migrate posts: code {:?}, err {}",
+                        code, err
+                    ))
+                })
+            });
+        let new_num_posts = read(|state| state.posts.len());
+        if new_num_posts == 0 {
+            // We're done with migration
+            break;
+        }
+        if num_posts == new_num_posts && num_posts > 0 {
+            // Something went wrong, so we stop and let the state locked.
+            mutate(|state| {
+                state
+                    .logger
+                    .error("couldn't migrate posts: something went wrong")
+            });
+            return;
+        }
+    }
+
+    if total_posts as u64 == Post::count() {
+        mutate(|state| {
+            state.read_only = false;
+            state
+                .logger
+                .debug(format!("Succefully migrated {} posts", total_posts))
+        })
+    }
+}
+
+#[update]
+fn migrate_posts() {
     mutate(|state| {
-        let num_posts = Post::count(state);
-        for id in 0..state.next_post_id {
-            Post::mutate(state, &id, |post| {
-                post.archived = false;
-                Ok(())
-            })
-            .unwrap();
+        let mut r = 0;
+        for _ in 0..20000 {
+            let Some((_, post)) = state.posts.pop_first() else {
+                break;
+            };
+            Post::save(post);
```

Comment:
Would it make set to `Post::get()` and compare with the old value to make sure that the post was saved correctly? If the comparison fails, then move the post to a new `state.failed_posts` map?

Reply:
Definitely!